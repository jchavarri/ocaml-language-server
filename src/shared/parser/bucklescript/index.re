external vscodeLangServerTypes : Js.t {..} = "vscode-languageserver-types" [@@bs.module];

/* TODO: add to glennsl/bs-vscode */
module DiagnosticSeverity = {
  type t;
  let error: t = vscodeLangServerTypes##_DiagnosticSeverity##_Error;
  let warning: t = vscodeLangServerTypes##_DiagnosticSeverity##_Warning;
  let information: t = vscodeLangServerTypes##_DiagnosticSeverity##_Information;
  let hint: t = vscodeLangServerTypes##_DiagnosticSeverity##_Hint;
};

type source =
  | Bucklescript
  | Merlin;

let sourceToString source =>
  switch source {
  | Bucklescript => "bucklescript"
  | Merlin => "merlin"
  };

type lineAndChar = {
  character: int,
  line: int
};

type range = {
  endPos: lineAndChar,
  startPos: lineAndChar
};

type diagnostic = {
  code: string,
  message: string,
  range,
  severity: DiagnosticSeverity.t,
  source: string
};

let getSeverity input => {
  let severityRe = [%re "/^Warning number \\d+/"];
  switch (severityRe |> Js.Re.exec input) {
  | Some _ => DiagnosticSeverity.warning
  | None => DiagnosticSeverity.error
  }
};

let createDiagnostic message startCharacter startLine endCharacter endLine severity => {
  "code": "",
  "message": message,
  "range": {
    "end": {character: endCharacter, line: endLine},
    "start": {character: startCharacter, line: startLine}
  },
  "severity": severity,
  "source": sourceToString Bucklescript
};

let unsafeGetMatch match =>
  switch (Js.Nullable.to_opt match) {
  | Some m => m
  | None => raise (Failure "Match not found")
  };

let parseErrors bsbOutput => {
  let parsedDiagnostics = Js.Dict.empty ();
  let level1 = [
    [%re "/File \"(.*)\", line (\\d*), characters (\\d*)-(\\d*):[\\s\\S]*?/"],
    [%re "/Error: ([\\s\\S]*)We've found a bug for you!/"]
  ];
  let reLevel1Errors =
    Js.Re.fromStringWithFlags (String.concat "" (List.map Js.Re.source level1)) flags::"g";
  let break = ref false;
  while (not !break) {
    switch (reLevel1Errors |> Js.Re.exec bsbOutput) {
    | None => break := true
    | Some result =>
      let matches = Js.Re.matches result;
      let fileUri = "file://" ^ unsafeGetMatch matches.(1);
      let startLine = int_of_string (unsafeGetMatch matches.(2)) - 1;
      let endLine = startLine;
      let startCharacter = int_of_string (unsafeGetMatch matches.(3));
      let endCharacter = int_of_string (unsafeGetMatch matches.(4));
      let message = String.trim (unsafeGetMatch matches.(5));
      let severity = getSeverity (unsafeGetMatch matches.(0));
      let diagnostic =
        createDiagnostic message startCharacter startLine endCharacter endLine severity;
      let diagnostics =
        switch (Js.Dict.get parsedDiagnostics fileUri) {
        | Some d => Js.Array.concat [|diagnostic|] d
        | None => [|diagnostic|]
        };
      Js.Dict.set parsedDiagnostics fileUri diagnostics
    }
  };
  let level2 = [
    [%re
      "/(?:We've found a bug for you!|Warning number \\d+)\\n\\s*/"
    ], /* Heading of the error / warning */
    [%re
      "/(.*) (\\d+):(\\d+)(?:-(\\d+)(?::(\\d+))?)?\\n  \\n/"
    ], /* Capturing file name and lines / indexes */
    [%re "/(?:.|\\n)*?\\n  \\n/"], /* Ignoring actual lines content being printed */
    [%re "/((?:.|\\n)*?)/"], /* Capturing error / warning message */
    [%re
      "/((?=We've found a bug for you!)|(?:\\[\\d+\\/\\d+\\] (?:\\x1b\\[[0-9;]*?m)?Building)|(?:ninja: build stopped: subcommand failed)|(?=Warning number \\d+)|$)/"
    ] /* Possible tails */
  ];
  let reLevel2Errors =
    Js.Re.fromStringWithFlags (String.concat "" (List.map Js.Re.source level2)) flags::"g";
  let break = ref false;
  while (not !break) {
    switch (reLevel2Errors |> Js.Re.exec bsbOutput) {
    | None => break := true
    | Some result =>
      let matches = Js.Re.matches result;
      let fileUri = "file://" ^ unsafeGetMatch matches.(1);
      let startLine = int_of_string (unsafeGetMatch matches.(2)) - 1;
      let startCharacter = int_of_string (unsafeGetMatch matches.(3)) - 1;
      let (endLine, endCharacter) =
        switch (Js.Nullable.to_opt matches.(4), Js.Nullable.to_opt matches.(5)) {
        /* Format is SL-SC:EL-EC */
        | (Some fourth, Some fifth) => (int_of_string fourth - 1, int_of_string fifth)
        /* Format is L:SC-EC */
        | (Some fourth, None) => (startLine, int_of_string fourth)
        /* Format is L:C */
        | (None, None) => (startLine, startCharacter)
        /* Impossible */
        | (None, Some _) => raise (Failure "Impossible state")
        };
      let messageRe = [%re "/\\n  /g"];
      let message = Js.String.replaceByRe messageRe "\\n" (unsafeGetMatch matches.(6));
      let severity = getSeverity (unsafeGetMatch matches.(0));
      let diagnostic =
        createDiagnostic message startCharacter startLine endCharacter endLine severity;
      let diagnostics =
        switch (Js.Dict.get parsedDiagnostics fileUri) {
        | Some d => Js.Array.concat [|diagnostic|] d
        | None => [|diagnostic|]
        };
      Js.Dict.set parsedDiagnostics fileUri diagnostics
    }
  };
  Js.log parsedDiagnostics;
  parsedDiagnostics
};
