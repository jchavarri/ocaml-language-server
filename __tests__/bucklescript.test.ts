import * as fs from "fs";
import * as path from "path";
import * as bucklescript from "../src/bin/server/parser/bucklescript";

const mocksFolder = "bucklescript.mocks";
const getDiagnostics = mockFileName => {
  const bsbOutput = fs.readFileSync(path.join(__dirname, mocksFolder, mockFileName), "utf8");
  return bucklescript.parseErrors(bsbOutput);
};

it('parses messages including "can\'t be found" strings', () => {
  expect(getDiagnostics("1.txt")).toMatchSnapshot();
});

it('parses messages including "Error while running external preprocessor"', () => {
  expect(getDiagnostics("2.txt")).toMatchSnapshot();
});

it('parses messages including "Error: 1087: Expecting one of the following"', () => {
  expect(getDiagnostics("3.txt")).toMatchSnapshot();
});

it("parses messages including multiple errors", () => {
  expect(getDiagnostics("4.txt")).toMatchSnapshot();
});

it('parses messages including "Unused bucklescript attribute"', () => {
  expect(getDiagnostics("5.txt")).toMatchSnapshot();
});

it("parses messages including interface/implementation mismatch", () => {
  expect(getDiagnostics("6.txt")).toMatchSnapshot();
});

it("parses messages including FFI errors with [@bs.send]", () => {
  expect(getDiagnostics("7.txt")).toMatchSnapshot();
});
