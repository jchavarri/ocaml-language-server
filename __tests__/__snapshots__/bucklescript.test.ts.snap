// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses messages including "Error while running external preprocessor" 1`] = `
Object {
  "file:///Users/bla/Development/sandbox/reason-react-hello/src/simple/bla.re": Array [
    Object {
      "code": "",
      "message": "2543: The switch expression isn't closed.",
      "range": Object {
        "end": Object {
          "character": 37,
          "line": 13,
        },
        "start": Object {
          "character": 36,
          "line": 13,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;

exports[`parses messages including "Error: 1087: Expecting one of the following" 1`] = `
Object {
  "file:///Users/bla/Development/github/the-super-tiny-compiler/src/compiler.re": Array [
    Object {
      "code": "",
      "message": "1087: Expecting one of the following:
  - \\"|\\" to open the next pattern
  - \\"=>\\" to start the body of the matched pattern
  - \\"when\\" to start a contitional guard for the previous pattern",
      "range": Object {
        "end": Object {
          "character": 14,
          "line": 497,
        },
        "start": Object {
          "character": 13,
          "line": 497,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;

exports[`parses messages including "Unused bucklescript attribute" 1`] = `
Object {
  "file:///Users/bla/Development/sandbox/my-first-app/src/Demo.re": Array [
    Object {
      "code": "",
      "message": "Unused bucklescript attribute: bs.as",
      "range": Object {
        "end": Object {
          "character": 11,
          "line": 6,
        },
        "start": Object {
          "character": 6,
          "line": 6,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;

exports[`parses messages including "can't be found" strings 1`] = `
Object {
  "file:///Users/bla/Development/sandbox/reason-hello/src/a.re": Array [
    Object {
      "code": "",
      "message": "  The value lfdsog can't be found in Js

",
      "range": Object {
        "end": Object {
          "character": 9,
          "line": 4,
        },
        "start": Object {
          "character": 0,
          "line": 4,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;

exports[`parses messages including FFI errors with [@bs.send] 1`] = `
Object {
  "file:///Users/bla/Development/sandbox/my-first-app/src/Demo.re": Array [
    Object {
      "code": "",
      "message": "You used an FFI attribute that can't be used with [@bs.send]",
      "range": Object {
        "end": Object {
          "character": 52,
          "line": 62,
        },
        "start": Object {
          "character": 0,
          "line": 62,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;

exports[`parses messages including interface/implementation mismatch 1`] = `
Object {
  "file:///Users/bla/Development/github/bs-bn.js/src/Bn.re": Array [
    Object {
      "code": "",
      "message": "The implementation src/Bn.mlast
     does not match the interface src/bn.cmi:
     The value \`toString' is required but not provided

",
      "range": Object {
        "end": Object {
          "character": 0,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;

exports[`parses messages including multiple errors 1`] = `
Object {
  "file:///Users/bla/Development/sandbox/reason-hello/src/a.re": Array [
    Object {
      "code": "",
      "message": "  Unbound value astsdfNode

Hint: Did you mean astNode?

",
      "range": Object {
        "end": Object {
          "character": 17,
          "line": 1,
        },
        "start": Object {
          "character": 7,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
  "file:///Users/bla/Development/sandbox/reason-hello/src/demo.re": Array [
    Object {
      "code": "",
      "message": "  Unbound value pattesdfrnMatch

Hint: Did you mean patternMatch?

",
      "range": Object {
        "end": Object {
          "character": 23,
          "line": 58,
        },
        "start": Object {
          "character": 8,
          "line": 58,
        },
      },
      "severity": 1,
      "source": "bucklescript",
    },
  ],
}
`;
